src/coreclr/vm/methodtable.cpp

```diff
@@ -7797,12 +7799,36 @@ MethodDesc* MethodTable::GetParallelMethodDesc(MethodDesc* pDefMD)
     }
     CONTRACTL_END;
 
+    if (asyncVariantLookup == AsyncVariantLookup::MatchingAsyncVariant)
+    {
 #ifdef FEATURE_METADATA_UPDATER
-    if (pDefMD->IsEnCAddedMethod())
-        return GetParallelMethodDescForEnC(this, pDefMD);
+        if (pDefMD->IsEnCAddedMethod())
+            return GetParallelMethodDescForEnC(this, pDefMD);
 #endif // FEATURE_METADATA_UPDATER
 
-    return GetMethodDescForSlot_NoThrow(pDefMD->GetSlot()); // TODO! We should probably use the throwing variant where possible
+        return GetMethodDescForSlot_NoThrow(pDefMD->GetSlot()); // TODO! We should probably use the throwing variant where possible
+    }
+    else
+    {
+        // Slow path for finding the Async variant (or not-Async variant, if we start from Async one)
+        // This could be optimized with some trickery around slot numbers, but doing so is ... confusing, so I'm not implementing this yet
```

Comment:
A solution for this was prototyped in https://github.com/dotnet/runtimelab/pull/3086 

It works, but it is a space-for-speed tradeoff of a kind that we typically do only for something that impacts run time (vs. JIT time), so we'd need more perf data to see if such change is a necessary or what we have is good enough.

Reply:
A less intrusive change for cases where both variants live in the same chunk is also possible. 

Either way it is better to do perf tuning when we can measure the impact on something closer to the realworld app. We may have other areas more likely to move the needle.